File: trading-algo/src/shared/messaging.ts
import type { Signal, StrategyDNA, PerformanceMetrics, Portfolio } from './types.js';
import { createModuleLogger } from './logger.js';
import { eventBus } from './events.js';

const log = createModuleLogger('messaging');

// ============================================================
// Message Types
// ============================================================

export type MessageChannel = 'claude' | 'openclaw';

export interface MessageTarget {
  channel: MessageChannel;
  /** OpenClaw webhook URL or API endpoint */
  endpoint?: string;
  /** OpenClaw API key for authentication */
  apiKey?: string;
  /** OpenClaw destination (e.g. 'telegram', 'discord', 'whatsapp') */
  destination?: string;
}

export type MessagePriority = 'low' | 'normal' | 'high' | 'critical';

export interface TradingMessage {
  id: string;
  type: 'evolution' | 'signal' | 'portfolio' | 'alert' | 'system';
  title: string;
  body: string;
  priority: MessagePriority;
  timestamp: number;
  metadata: Record<string, unknown>;
}

// ============================================================
// Message Formatters
// ============================================================

export function formatEvolutionMessage(data: {
  strategy: string;
  generation: number;
  fitness: number;
  previousFitness: number;
  params: Record<string, number>;
  improved: boolean;
}): TradingMessage {
  const improvement = data.previousFitness > 0
    ? ((data.fitness - data.previousFitness) / data.previousFitness * 100).toFixed(1)
    : 'N/A';

  const paramsSummary = Object.entries(data.params)
    .map(([k, v]) => `  ${k}: ${typeof v === 'number' ? v.toFixed(4) : v}`)
    .join('\n');

  const emoji = data.improved ? '[IMPROVED]' : '[STABLE]';

  return {
    id: `evo-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
    type: 'evolution',
    title: `${emoji} Strategy Evolution: ${data.strategy}`,
    body: [
      `Strategy: ${data.strategy}`,
      `Generation: ${data.generation}`,
      `Fitness: ${data.fitness.toFixed(4)} (${data.improved ? `+${improvement}%` : 'no improvement'})`,
      `Previous Fitness: ${data.previousFitness.toFixed(4)}`,
      '',
      'Parameters:',
      paramsSummary,
    ].join('\n'),
    priority: data.improved ? 'high' : 'normal',
    timestamp: Date.now(),
    metadata: { ...data },
  };
}

export function formatSignalMessage(signal: Signal): TradingMessage {
  const priority: MessagePriority = signal.confidence > 0.8 ? 'high'
    : signal.confidence > 0.6 ? 'normal' : 'low';

  const indicatorsSummary = Object.entries(signal.indicators)
    .map(([k, v]) => `  ${k}: ${typeof v === 'number' ? v.toFixed(4) : v}`)
    .join('\n');

  return {
    id: `sig-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
    type: 'signal',
    title: `[${signal.action}] ${signal.asset.symbol} @ ${signal.price.toFixed(5)}`,
    body: [
      `Asset: ${signal.asset.symbol} (${signal.asset.assetClass})`,
      `Action: ${signal.action}`,
      `Price: ${signal.price.toFixed(5)}`,
      `Confidence: ${(signal.confidence * 100).toFixed(1)}%`,
      `Strategy: ${signal.strategy}`,
      `Timeframe: ${signal.timeframe}`,
      `Reason: ${signal.reason}`,
      '',
      'Indicators:',
      indicatorsSummary,
    ].join('\n'),
    priority,
    timestamp: signal.timestamp,
    metadata: { signal },
  };
}

export function formatPortfolioMessage(portfolio: Portfolio): TradingMessage {
  const positionsSummary = portfolio.positions
    .filter(p => p.status === 'open')
    .map(p => `  ${p.asset.symbol}: ${p.side} ${p.quantity} @ ${p.entryPrice.toFixed(5)} (PnL: ${p.unrealizedPnl.toFixed(2)})`)
    .join('\n') || '  (no open positions)';

  return {
    id: `port-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
    type: 'portfolio',
    title: `Portfolio Update`,
    body: [
      `Capital: $${portfolio.capital.toFixed(2)}`,
      `Available: $${portfolio.availableCapital.toFixed(2)}`,
      `Total PnL: $${portfolio.totalPnl.toFixed(2)} (${portfolio.totalPnlPct.toFixed(2)}%)`,
      `Max Drawdown: ${portfolio.maxDrawdown.toFixed(2)}%`,
      `Open Positions: ${portfolio.positions.filter(p => p.status === 'open').length}`,
      '',
      'Positions:',
      positionsSummary,
    ].join('\n'),
    priority: Math.abs(portfolio.totalPnlPct) > 5 ? 'high' : 'normal',
    timestamp: portfolio.lastUpdated,
    metadata: { portfolio },
  };
}

export function formatAlertMessage(
  title: string,
  body: string,
  priority: MessagePriority = 'high',
): TradingMessage {
  return {
    id: `alert-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
    type: 'alert',
    title,
    body,
    priority,
    timestamp: Date.now(),
    metadata: {},
  };
}

// ============================================================
// Message Dispatcher
// ============================================================

/**
 * Dispatches trading messages to configured channels.
 * Supports Claude Code (stdout/log) and OpenClaw (webhook).
 */
export class MessageDispatcher {
  private targets: MessageTarget[] = [];
  private messageHistory: TradingMessage[] = [];
  private maxHistory = 100;

  constructor(targets?: MessageTarget[]) {
    if (targets) {
      this.targets = targets;
    } else {
      // Default: Claude Code output
      this.targets = [{ channel: 'claude' }];

      // Auto-detect OpenClaw from env
      const openclawEndpoint = process.env.OPENCLAW_WEBHOOK_URL ?? process.env.OPENCLAW_ENDPOINT;
      if (openclawEndpoint) {
        this.targets.push({
          channel: 'openclaw',
          endpoint: openclawEndpoint,
          apiKey: process.env.OPENCLAW_API_KEY,
          destination: process.env.OPENCLAW_DESTINATION ?? 'telegram',
        });
      }
    }

    log.info({ channels: this.targets.map(t => t.channel) }, 'MessageDispatcher initialized');
  }

  addTarget(target: MessageTarget): void {
    this.targets.push(target);
    log.info({ channel: target.channel }, 'Message target added');
  }

  removeTarget(channel: MessageChannel): void {
    this.targets = this.targets.filter(t => t.channel !== channel);
  }

  getTargets(): MessageTarget[] {
    return [...this.targets];
  }

  /**
   * Send a message to all configured targets.
   */
  async send(message: TradingMessage): Promise<void> {
    this.messageHistory.push(message);
    if (this.messageHistory.length > this.maxHistory) {
      this.messageHistory = this.messageHistory.slice(-this.maxHistory);
    }

    for (const target of this.targets) {
      try {
        switch (target.channel) {
          case 'claude':
            this.sendToClaude(message);
            break;
          case 'openclaw':
            await this.sendToOpenClaw(message, target);
            break;
        }
      } catch (err) {
        log.error({ channel: target.channel, error: err }, 'Failed to send message');
      }
    }
  }

  /**
   * Send to Claude Code (structured log output).
   */
  private sendToClaude(message: TradingMessage): void {
    const logFn = message.priority === 'critical' || message.priority === 'high'
      ? log.warn.bind(log) : log.info.bind(log);

    logFn({
      messageType: message.type,
      title: message.title,
      priority: message.priority,
    }, `[MSG] ${message.title}\n${message.body}`);
  }

  /**
   * Send to OpenClaw via webhook/API.
   * OpenClaw expects a JSON payload with message content and routing info.
   */
  private async sendToOpenClaw(message: TradingMessage, target: MessageTarget): Promise<void> {
    if (!target.endpoint) {
      log.warn('OpenClaw endpoint not configured — skipping');
      return;
    }

    const payload = {
      // OpenClaw standard fields
      message: `**${message.title}**\n\n${message.body}`,
      destination: target.destination ?? 'telegram',
      priority: message.priority,
      // Structured data for OpenClaw agents
      structured: {
        id: message.id,
        type: message.type,
        title: message.title,
        body: message.body,
        metadata: message.metadata,
        timestamp: message.timestamp,
      },
    };

    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };
    if (target.apiKey) {
      headers['Authorization'] = `Bearer ${target.apiKey}`;
    }

    try {
      const response = await fetch(target.endpoint, {
        method: 'POST',
        headers,
        body: JSON.stringify(payload),
        signal: AbortSignal.timeout(5000),
      });

      if (!response.ok) {
        log.warn({
          status: response.status,
          endpoint: target.endpoint,
        }, 'OpenClaw webhook returned non-OK status');
      } else {
        log.debug({ endpoint: target.endpoint }, 'Message sent to OpenClaw');
      }
    } catch (err) {
      log.error({ error: err, endpoint: target.endpoint }, 'OpenClaw webhook failed');
    }
  }

  /**
   * Get recent message history.
   */
  getHistory(limit = 20): TradingMessage[] {
    return this.messageHistory.slice(-limit);
  }
}

// ============================================================
// Singleton + Event Wiring
// ============================================================

/** Global message dispatcher instance */
export const messageDispatcher = new MessageDispatcher();

/**
 * Wire up event bus to automatically dispatch messages for key events.
 * Call this once during system initialization.
 */
export function wireMessagingEvents(): void {
  eventBus.on('evolution:improvement', (event) => {
    const data = event.data as {
      strategy: string;
      generation: number;
      fitness: number;
      params: Record<string, number>;
    };
    const msg = formatEvolutionMessage({
      ...data,
      previousFitness: 0,
      improved: true,
    });
    messageDispatcher.send(msg);
  });

  eventBus.on('signal:generated', (event) => {
    const signal = event.data as Signal;
    // Only dispatch high-confidence signals to avoid noise
    if (signal.confidence >= 0.6) {
      const msg = formatSignalMessage(signal);
      messageDispatcher.send(msg);
    }
  });

  eventBus.on('risk:alert', (event) => {
    const data = event.data as { title: string; message: string };
    const msg = formatAlertMessage(
      data.title ?? 'Risk Alert',
      data.message ?? JSON.stringify(data),
      'critical',
    );
    messageDispatcher.send(msg);
  });

  log.info('Messaging events wired to EventBus');
}

File: trading-algo-py/shared/messaging.py
"""Message dispatching to multiple channels: Claude, KimiClaw, Ollama.

Ported from trading-algo/src/shared/messaging.ts.

Supports dual-model routing:
  - Claude Code (structured logs -- default)
  - KimiClaw (webhook -> WeChat, Telegram, Discord, WhatsApp)
  - Ollama / open-source LLMs (local API for AI-powered analysis)
"""

from __future__ import annotations

import json
import logging
import os
import time
import uuid
from dataclasses import dataclass, field
from typing import Any, Literal

import httpx

from shared.types import Signal, Portfolio, PerformanceMetrics, StrategyDNA
from shared.events import event_bus

log = logging.getLogger(__name__)

# ============================================================
# Message Types
# ============================================================

MessageChannel = Literal["claude", "kimiclaw", "ollama"]

MessageType = Literal["evolution", "signal", "portfolio", "alert", "system"]

MessagePriority = Literal["low", "normal", "high", "critical"]


@dataclass
class MessageTarget:
    channel: MessageChannel
    endpoint: str | None = None
    api_key: str | None = None
    destination: str | None = None
    model: str | None = None
    filter: list[MessageType] | None = None


@dataclass
class TradingMessage:
    id: str
    type: MessageType
    title: str
    body: str
    priority: MessagePriority
    timestamp: int
    metadata: dict[str, Any] = field(default_factory=dict)


# ============================================================
# Message Formatters
# ============================================================

def _generate_id(prefix: str) -> str:
    short = uuid.uuid4().hex[:6]
    return f"{prefix}-{int(time.time() * 1000)}-{short}"


def format_evolution_message(
    *,
    strategy: str,
    generation: int,
    fitness: float,
    previous_fitness: float,
    params: dict[str, float],
    improved: bool,
) -> TradingMessage:
    if previous_fitness > 0:
        improvement = f"{(fitness - previous_fitness) / previous_fitness * 100:.1f}"
    else:
        improvement = "N/A"

    params_summary = "\n".join(
        f"  {k}: {v:.4f}" if isinstance(v, (int, float)) else f"  {k}: {v}"
        for k, v in params.items()
    )

    tag = "[IMPROVED]" if improved else "[STABLE]"

    return TradingMessage(
        id=_generate_id("evo"),
        type="evolution",
        title=f"{tag} Strategy Evolution: {strategy}",
        body="\n".join([
            f"Strategy: {strategy}",
            f"Generation: {generation}",
            f"Fitness: {fitness:.4f} ({f'+{improvement}%' if improved else 'no improvement'})",
            f"Previous Fitness: {previous_fitness:.4f}",
            "",
            "Parameters:",
            params_summary,
        ]),
        priority="high" if improved else "normal",
        timestamp=int(time.time() * 1000),
        metadata={
            "strategy": strategy,
            "generation": generation,
            "fitness": fitness,
            "previous_fitness": previous_fitness,
            "params": params,
            "improved": improved,
        },
    )


def format_signal_message(signal: Signal) -> TradingMessage:
    if signal.confidence > 0.8:
        priority: MessagePriority = "high"
    elif signal.confidence > 0.6:
        priority = "normal"
    else:
        priority = "low"

    indicators_summary = "\n".join(
        f"  {k}: {v:.4f}" if isinstance(v, (int, float)) else f"  {k}: {v}"
        for k, v in signal.indicators.items()
    )

    return TradingMessage(
        id=_generate_id("sig"),
        type="signal",
        title=f"[{signal.action.value}] {signal.asset.symbol} @ {signal.price:.5f}",
        body="\n".join([
            f"Asset: {signal.asset.symbol} ({signal.asset.asset_class.value})",
            f"Action: {signal.action.value}",
            f"Price: {signal.price:.5f}",
            f"Confidence: {signal.confidence * 100:.1f}%",
            f"Strategy: {signal.strategy}",
            f"Timeframe: {signal.timeframe}",
            f"Reason: {signal.reason}",
            "",
            "Indicators:",
            indicators_summary,
        ]),
        priority=priority,
        timestamp=signal.timestamp,
        metadata={"signal": signal.__dict__},
    )


def format_portfolio_message(portfolio: Portfolio) -> TradingMessage:
    open_positions = [p for p in portfolio.positions if p.status.value == "open"]
    if open_positions:
        positions_summary = "\n".join(
            f"  {p.asset.symbol}: {p.side.value} {p.quantity} @ {p.entry_price:.5f}"
            f" (PnL: {p.unrealized_pnl:.2f})"
            for p in open_positions
        )
    else:
        positions_summary = "  (no open positions)"

    return TradingMessage(
        id=_generate_id("port"),
        type="portfolio",
        title="Portfolio Update",
        body="\n".join([
            f"Capital: ${portfolio.capital:.2f}",
            f"Available: ${portfolio.available_capital:.2f}",
            f"Total PnL: ${portfolio.total_pnl:.2f} ({portfolio.total_pnl_pct:.2f}%)",
            f"Max Drawdown: {portfolio.max_drawdown:.2f}%",
            f"Open Positions: {len(open_positions)}",
            "",
            "Positions:",
            positions_summary,
        ]),
        priority="high" if abs(portfolio.total_pnl_pct) > 5 else "normal",
        timestamp=portfolio.last_updated,
        metadata={"portfolio": portfolio.__dict__},
    )


def format_alert_message(
    title: str,
    body: str,
    priority: MessagePriority = "high",
) -> TradingMessage:
    return TradingMessage(
        id=_generate_id("alert"),
        type="alert",
        title=title,
        body=body,
        priority=priority,
        timestamp=int(time.time() * 1000),
        metadata={},
    )


# ============================================================
# Message Dispatcher
# ============================================================

class MessageDispatcher:
    """Dispatches trading messages to configured channels.

    Supports dual-model routing:
      - Claude Code (structured logs -- default)
      - KimiClaw (webhook -> WeChat, Telegram, Discord, WhatsApp)
      - Ollama / open-source LLMs (local API for AI-powered analysis)

    Configure via env vars:
      KIMICLAW_WEBHOOK_URL, KIMICLAW_API_KEY, KIMICLAW_DESTINATION, KIMICLAW_MODEL
      OLLAMA_ENDPOINT (default: http://localhost:11434), OLLAMA_MODEL (default: llama3)
    """

    def __init__(self, targets: list[MessageTarget] | None = None) -> None:
        self._targets: list[MessageTarget] = []
        self._message_history: list[TradingMessage] = []
        self._max_history = 100

        if targets is not None:
            self._targets = list(targets)
        else:
            # Default: Claude Code output
            self._targets.append(MessageTarget(channel="claude"))

            # Auto-detect KimiClaw from env
            kimiclaw_endpoint = (
                os.environ.get("KIMICLAW_WEBHOOK_URL")
                or os.environ.get("KIMICLAW_ENDPOINT")
            )
            if kimiclaw_endpoint:
                self._targets.append(MessageTarget(
                    channel="kimiclaw",
                    endpoint=kimiclaw_endpoint,
                    api_key=os.environ.get("KIMICLAW_API_KEY"),
                    destination=os.environ.get("KIMICLAW_DESTINATION", "telegram"),
                    model=os.environ.get("KIMICLAW_MODEL", "kimi"),
                ))

            # Auto-detect Ollama (local open-source LLM)
            ollama_endpoint = os.environ.get("OLLAMA_ENDPOINT")
            if ollama_endpoint:
                self._targets.append(MessageTarget(
                    channel="ollama",
                    endpoint=ollama_endpoint,
                    model=os.environ.get("OLLAMA_MODEL", "llama3"),
                    # Ollama only gets evolution + alert + portfolio messages by default
                    filter=["evolution", "alert", "portfolio"],
                ))

        channels = [
            f"{t.channel}({t.model})" if t.model else t.channel
            for t in self._targets
        ]
        log.info("MessageDispatcher initialized: channels=%s", channels)

    def add_target(self, target: MessageTarget) -> None:
        self._targets.append(target)
        log.info("Message target added: channel=%s model=%s", target.channel, target.model)

    def remove_target(self, channel: MessageChannel) -> None:
        self._targets = [t for t in self._targets if t.channel != channel]

    def get_targets(self) -> list[MessageTarget]:
        return list(self._targets)

    async def send(self, message: TradingMessage) -> None:
        """Send a message to all configured targets.

        Each target can filter by message type.
        """
        self._message_history.append(message)
        if len(self._message_history) > self._max_history:
            self._message_history = self._message_history[-self._max_history :]

        for target in self._targets:
            # Check filter -- skip if target doesn't want this message type
            if target.filter and message.type not in target.filter:
                continue

            try:
                if target.channel == "claude":
                    self._send_to_claude(message)
                elif target.channel == "kimiclaw":
                    await self._send_to_kimiclaw(message, target)
                elif target.channel == "ollama":
                    await self._send_to_ollama(message, target)
            except Exception:
                log.error(
                    "Failed to send message to channel=%s",
                    target.channel,
                    exc_info=True,
                )

    def _send_to_claude(self, message: TradingMessage) -> None:
        """Send to Claude Code (structured log output)."""
        log_fn = (
            log.warning
            if message.priority in ("critical", "high")
            else log.info
        )
        log_fn(
            "[MSG] %s\n%s",
            message.title,
            message.body,
            extra={"messageType": message.type, "priority": message.priority},
        )

    async def _send_to_kimiclaw(
        self, message: TradingMessage, target: MessageTarget,
    ) -> None:
        """Send to KimiClaw via webhook/API.

        KimiClaw routes messages to WeChat, Telegram, Discord, WhatsApp, etc.
        """
        if not target.endpoint:
            log.warning("KimiClaw endpoint not configured -- skipping")
            return

        payload = {
            "message": f"**{message.title}**\n\n{message.body}",
            "destination": target.destination or "telegram",
            "model": target.model or "kimi",
            "priority": message.priority,
            "structured": {
                "id": message.id,
                "type": message.type,
                "title": message.title,
                "body": message.body,
                "metadata": message.metadata,
                "timestamp": message.timestamp,
            },
        }

        await self._post_webhook(target, payload, "KimiClaw")

    async def _send_to_ollama(
        self, message: TradingMessage, target: MessageTarget,
    ) -> None:
        """Send to a local Ollama instance for AI-powered analysis.

        The open-source LLM can analyze evolution results, suggest parameter
        tweaks, or provide a second opinion on trading signals.
        """
        if not target.endpoint:
            log.warning("Ollama endpoint not configured -- skipping")
            return

        system_prompt = (
            "You are a quantitative trading analyst. Analyze the following "
            "trading system update and provide brief, actionable insights. "
            "Focus on risk, opportunity, and what to monitor next."
        )

        payload = {
            "model": target.model or "llama3",
            "messages": [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": f"{message.title}\n\n{message.body}"},
            ],
            "stream": False,
        }

        endpoint = target.endpoint.rstrip("/") + "/api/chat"

        try:
            async with httpx.AsyncClient() as client:
                response = await client.post(
                    endpoint,
                    json=payload,
                    timeout=30.0,  # LLM inference can be slow
                )

            if response.status_code == 200:
                result = response.json()
                analysis = result.get("message", {}).get("content", "")
                if analysis:
                    log.info(
                        "[Ollama/%s] %s",
                        target.model,
                        analysis,
                        extra={
                            "model": target.model,
                            "messageType": message.type,
                            "analysis": analysis[:200],
                        },
                    )
            else:
                log.warning(
                    "Ollama returned non-OK status: %d endpoint=%s",
                    response.status_code,
                    endpoint,
                )
        except Exception:
            log.error("Ollama request failed: endpoint=%s", endpoint, exc_info=True)

    async def _post_webhook(
        self,
        target: MessageTarget,
        payload: dict[str, Any],
        label: str,
    ) -> None:
        """Generic webhook POST helper."""
        headers: dict[str, str] = {"Content-Type": "application/json"}
        if target.api_key:
            headers["Authorization"] = f"Bearer {target.api_key}"

        try:
            async with httpx.AsyncClient() as client:
                response = await client.post(
                    target.endpoint,
                    json=payload,
                    headers=headers,
                    timeout=5.0,
                )

            if response.status_code >= 400:
                log.warning(
                    "%s webhook returned non-OK status: %d endpoint=%s",
                    label,
                    response.status_code,
                    target.endpoint,
                )
            else:
                log.debug("Message sent to %s endpoint=%s", label, target.endpoint)
        except Exception:
            log.error(
                "%s webhook failed: endpoint=%s", label, target.endpoint, exc_info=True,
            )

    def get_history(self, limit: int = 20) -> list[TradingMessage]:
        """Get recent message history."""
        return self._message_history[-limit:]


# ============================================================
# Singleton + Event Wiring
# ============================================================

message_dispatcher = MessageDispatcher()


def wire_messaging_events() -> None:
    """Wire up event bus to automatically dispatch messages for key events.

    Call this once during system initialization.
    """

    def _on_evolution_improvement(data: Any, source: str = "") -> None:
        msg = format_evolution_message(
            strategy=data.get("strategy", ""),
            generation=data.get("generation", 0),
            fitness=data.get("fitness", 0.0),
            previous_fitness=0.0,
            params=data.get("params", {}),
            improved=True,
        )
        # fire-and-forget: use asyncio if available
        import asyncio

        try:
            loop = asyncio.get_running_loop()
            loop.create_task(message_dispatcher.send(msg))
        except RuntimeError:
            asyncio.run(message_dispatcher.send(msg))

    def _on_signal_generated(data: Any, source: str = "") -> None:
        signal: Signal = data
        # Only dispatch high-confidence signals to avoid noise
        if signal.confidence < 0.6:
            return
        msg = format_signal_message(signal)
        import asyncio

        try:
            loop = asyncio.get_running_loop()
            loop.create_task(message_dispatcher.send(msg))
        except RuntimeError:
            asyncio.run(message_dispatcher.send(msg))

    def _on_risk_alert(data: Any, source: str = "") -> None:
        msg = format_alert_message(
            title=data.get("title", "Risk Alert") if isinstance(data, dict) else "Risk Alert",
            body=data.get("message", json.dumps(data)) if isinstance(data, dict) else str(data),
            priority="critical",
        )
        import asyncio

        try:
            loop = asyncio.get_running_loop()
            loop.create_task(message_dispatcher.send(msg))
        except RuntimeError:
            asyncio.run(message_dispatcher.send(msg))

    event_bus.on("evolution:improvement", _on_evolution_improvement)
    event_bus.on("signal:generated", _on_signal_generated)
    event_bus.on("risk:alert", _on_risk_alert)

    log.info("Messaging events wired to EventBus")
